### ISSUE TITLE
Handle combi-soaks across the whole pipeline (XCA, loader, stack)

### ISSUE THREAD

**Issue by mwinokan** [m2ms/fragalysis-frontend#1432]
@ConorFWild says that combi-soaks are handled already in XCA and have been tested with Jasmin's dataset (DENV) but @tdudgeon says that the CIF parsing in the collator and alignment post-processing (using the CIF) will likely not handle multiple ligands. Multiple observations each with a different ligand are likely fine, but observations containing multiple ligands are not supported.

**Comment by mwinokan** [comment]
To test this take any observation with multiple ligands in a crystal (don't have to be chemically different) and change one of the ligand residue names to `XXX`. E.g. from Jasmin's (DENV):

@tdudgeon please check if this is enough to test: `/dls/labxchem/data/lb32633/lb32633-1/processing/analysis/xchemalign/upload_1/aligned_files/DNV2_NS5A-x0160`

**Comment by mwinokan** [comment]
@tdudgeon please see the above comment for some test data

**Comment by tdudgeon** [comment]
That crystal doesn't seem to have been processed as I would have expected. The crystal contains 4 copies of the same ligand. The CIF file seems to describe the ligand correctly (it's named LIG in the CIF and PDB files). The ligand definmition in the PDB is a little strange:
```
HETATM 5567  N   LIG A1608     -13.733 -40.017 -22.386  0.49 28.82           N
HETATM 5568  C   LIG A1608     -15.315 -43.378 -22.254  0.49 29.34           C
HETATM 5569  C1  LIG A1608     -14.839 -41.988 -22.585  0.49 29.23           C
HETATM 5570  C2  LIG A1608     -13.965 -41.198 -21.819  0.49 29.00           C
HETATM 5571  C3  LIG A1608     -15.144 -41.194 -23.702  0.49 28.44           C
HETATM 5572  C4  LIG A1608     -16.013 -41.420 -24.877  0.49 28.56           C
HETATM 5573  C5  LIG A1608     -15.933 -40.597 -25.993  0.49 28.13           C
HETATM 5574  C6  LIG A1608     -16.789 -40.826 -27.052  0.49 28.57           C
HETATM 5575  C7  LIG A1608     -17.756 -42.602 -26.001  0.49 29.75           C
HETATM 5576  C8  LIG A1608     -16.950 -42.448 -24.889  0.49 29.33           C
HETATM 5577  N1  LIG A1608     -14.464 -40.036 -23.538  0.49 28.79           N
HETATM 5578  N2  LIG A1608     -17.697 -41.808 -27.076  0.49 29.84           N
HETATM 5615  N  ALIG A1701     -41.077 -34.420 -26.864  0.30 33.79           N
HETATM 5616  C  ALIG A1701     -39.814 -30.998 -27.579  0.30 34.80           C
HETATM 5617  C1 ALIG A1701     -40.057 -32.477 -27.440  0.30 35.64           C
HETATM 5618  C2 ALIG A1701     -41.155 -33.094 -26.813  0.30 33.65           C
HETATM 5619  C3 ALIG A1701     -39.271 -33.543 -27.899  0.30 35.46           C
HETATM 5620  C4 ALIG A1701     -37.986 -33.583 -28.625  0.30 36.34           C
HETATM 5621  C5 ALIG A1701     -37.455 -32.438 -29.208  0.30 35.19           C
HETATM 5622  C6 ALIG A1701     -36.216 -32.513 -29.812  0.30 35.60           C
HETATM 5623  C7 ALIG A1701     -36.023 -34.732 -29.334  0.30 39.68           C
HETATM 5624  C8 ALIG A1701     -37.252 -34.759 -28.707  0.30 38.09           C
HETATM 5625  N1 ALIG A1701     -39.919 -34.674 -27.535  0.30 35.30           N
HETATM 5626  N2 ALIG A1701     -35.491 -33.633 -29.884  0.30 38.61           N
HETATM 5627  N  BLIG A1701     -40.422 -31.799 -26.902  0.26 52.37           N
HETATM 5628  C  BLIG A1701     -40.688 -35.488 -27.256  0.26 48.48           C
HETATM 5629  C1 BLIG A1701     -40.295 -34.035 -27.261  0.26 51.14           C
HETATM 5630  C2 BLIG A1701     -41.040 -32.965 -26.737  0.26 51.01           C
HETATM 5631  C3 BLIG A1701     -39.137 -33.438 -27.778  0.26 52.54           C
HETATM 5632  C4 BLIG A1701     -37.950 -33.998 -28.458  0.26 54.76           C
HETATM 5633  C5 BLIG A1701     -37.090 -33.181 -29.184  0.26 55.30           C
HETATM 5634  C6 BLIG A1701     -35.978 -33.746 -29.777  0.26 57.76           C
HETATM 5635  C7 BLIG A1701     -36.514 -35.824 -29.006  0.26 56.99           C
HETATM 5636  C8 BLIG A1701     -37.647 -35.353 -28.372  0.26 56.02           C
HETATM 5637  N1 BLIG A1701     -39.259 -32.112 -27.540  0.26 52.39           N
HETATM 5638  N2 BLIG A1701     -35.676 -35.047 -29.702  0.26 59.66           N
HETATM 5639  N   LIG A1702     -17.571 -40.981 -20.388  0.49 50.52           N
HETATM 5640  C   LIG A1702     -16.821 -37.430 -21.175  0.49 49.33           C
HETATM 5641  C1  LIG A1702     -17.281 -38.860 -21.138  0.49 49.37           C
HETATM 5642  C2  LIG A1702     -17.042 -39.793 -20.114  0.49 50.28           C
HETATM 5643  C3  LIG A1702     -18.019 -39.567 -22.093  0.49 49.38           C
HETATM 5644  C4  LIG A1702     -18.596 -39.171 -23.395  0.49 48.80           C
HETATM 5645  C5  LIG A1702     -18.031 -38.149 -24.149  0.49 48.20           C
HETATM 5646  C6  LIG A1702     -18.604 -37.825 -25.365  0.49 47.39           C
HETATM 5647  C7  LIG A1702     -20.208 -39.424 -25.135  0.49 48.12           C
HETATM 5648  C8  LIG A1702     -19.713 -39.820 -23.906  0.49 47.90           C
HETATM 5649  N1  LIG A1702     -18.163 -40.821 -21.606  0.49 50.24           N
HETATM 5650  N2  LIG A1702     -19.676 -38.442 -25.870  0.49 46.32           N
```
The second and third ligands seem to be alternatives (residue 1702) whilst the first and fourth are distinct (residues 1608 and 1702).

Only 3 alignments (not 4) are generated and they are essentially identical. It's not easy to determine how many sites are observed, it could be 2, 3 or 4. But if it really is 3 then 2 of them at least must be in quite a different location so you would expect to see at least 2 significantly different alignments.

@ConorFWild can you comment on this?

The crystallographic and aligned PDBs are:

[DNV2_NS5A-x0160.pdb.gz](https://github.com/m2ms/fragalysis-frontend/files/15390957/DNV2_NS5A-x0160.pdb.gz)
[DNV2_NS5A-x0160_A_1608_1_DNV2_NS5A-x0191+A+1608+1.pdb.gz](https://github.com/m2ms/fragalysis-frontend/files/15390958/DNV2_NS5A-x0160_A_1608_1_DNV2_NS5A-x0191%2BA%2B1608%2B1.pdb.gz)
[DNV2_NS5A-x0160_A_1701_1_DNV2_NS5A-x0160+A+1701+1.pdb.gz](https://github.com/m2ms/fragalysis-frontend/files/15390959/DNV2_NS5A-x0160_A_1701_1_DNV2_NS5A-x0160%2BA%2B1701%2B1.pdb.gz)
[DNV2_NS5A-x0160_A_1702_1_DNV2_NS5A-x0191+A+1608+1.pdb.gz](https://github.com/m2ms/fragalysis-frontend/files/15390960/DNV2_NS5A-x0160_A_1702_1_DNV2_NS5A-x0191%2BA%2B1608%2B1.pdb.gz)

Also, this crystal does not really test what we were wanting to test as the mulitple ligands are all the same molecule. We really need a crystal with 2 (or more) different ligands. @mwinokan,. @Waztom  do you know of any?

**Comment by mwinokan** [comment]
@ConorFWild does not know of any non-archived data with multiple ligands. **This might be a breaking change though, so we should find some test data soon** @mwinokan to have a look

**Comment by mwinokan** [comment]
- The DB model supports combi-soaks
- The loader does not support combi-soaks
- XCA does not support combi-soaks

@tdudgeon still needs test data (@mwinokan to get from Jasmin)

There's no point faking the data as the representation in SoakDB is important too

**Comment by mwinokan** [comment]
@tdudgeon I have spent a little time chatting to Jasmin and looking at her combo-soak data.

You can see an example in `/dls/labxchem/data/lb32633/lb32633-6/processing/analysis/model_building/CHIKV_MacB-x1669`

The things to note:

- In this directory there are 3 .cif files, one for each ligand observed in the dataset.
- The names of the ligand/residue within the .cif files is more or less arbitrary (chosen by the curator) but unique within the dataset
- The residue name of each ligand in the refine.pdb corresponds to one of the names in the .cif files.

E.g. for CIKV_MacB-x1669 there is `LIG`, `113` and `131`. 

As I see it, XCA should look for the names in the .cif's and use that to make sense of the PDB's for alignment. 

Other datasets for this target that have been (preliminarily) modelled with multiple ligands are:

- CHIKV_MacB-x1669.pdb
- CHIKV_MacB-x1674.pdb
- CHIKV_MacB-x1678.pdb
- CHIKV_MacB-x1689.pdb
- CHIKV_MacB-x1692.pdb
- CHIKV_MacB-x1704.pdb
- CHIKV_MacB-x1734.pdb
- CHIKV_MacB-x1739.pdb
- CHIKV_MacB-x1740.pdb
- CHIKV_MacB-x1742.pdb

Is this enough to start working with?

**Comment by tdudgeon** [comment]
@mwinokan So here lies the problem. The data in soakdb is used the identify the correct CIF to use (as well as the right PDB, MTZ etc.). It's the value in the `RefinementCIF` column, and that just has a single value (`/dls/labxchem/data/lb32633/lb32633-6/processing/analysis/model_building/CHIKV_MacB-x1669/compound/Z1041785508.cif`). That indeed is one of the 3 CIF files you mention, but AFAICT soakdb is not aware of the other 2 CIF files.



</body></html>

**Comment by mwinokan** [comment]
@tdudgeon could you please share the entire SoakDB entry for `CHIKV_MacB-x1669`.

Is it not safe to assume that every `*.cif` in the corresponding model directory is present in the PDBs?

**Comment by tdudgeon** [comment]
```
ID | LabVisit | LibraryPlate | SourceWell | LibraryName | CompoundSMILES | CompoundCode | CrystalPlate | CrystalWell | EchoX | EchoY | DropVolume | ProteinName | BatchNumber | CompoundStockConcentration | CompoundConcentration | SolventFraction | SoakTransferVol | SoakStatus | SoakTimestamp | CryoStockFraction | CryoFraction | CryoWell | CryoTransferVolume | CryoStatus | CryoTimestamp | SoakingTime | HarvestStatus | CrystalName | Puck | PuckPosition | PinBarcode | MountingResult | MountingArrivalTime | MountedTimestamp | MountingTime | ispybStatus | DataCollectionVisit | SoakDBComments | CompoundSMILESproduct | CompoundStereo | CompoundStereoSMILES | CompoundStereoCIFprogram | CompoundStereoCIFs | CompoundAutofitCC | CompoundAutofitprogram | ProjectDirectory | CrystalTag | CrystalFormName | CrystalFormSpaceGroup | CrystalFormPointGroup | CrystalFormA | CrystalFormB | CrystalFormC | CrystalFormAlpha | CrystalFormBeta | CrystalFormGamma | CrystalFormVolume | DataCollectionBeamline | DataCollectionDate | DataCollectionOutcome | DataCollectionRun | DataCollectionSubdir | DataCollectionComment | DataCollectionWavelength | DataCollectionPinBarcode | DataCollectionCrystalImage1 | DataCollectionCrystalImage2 | DataCollectionCrystalImage3 | DataCollectionCrystalImage4 | DataProcessingPathToImageFiles | DataProcessingProgram | DataProcessingSpaceGroup | DataProcessingUnitCell | DataProcessingAutoAssigned | DataProcessingA | DataProcessingB | DataProcessingC | DataProcessingAlpha | DataProcessingBeta | DataProcessingGamma | DataProcessingResolutionOverall | DataProcessingResolutionLow | DataProcessingResolutionLowInnerShell | DataProcessingResolutionHigh | DataProcessingResolutionHigh15sigma | DataProcessingResolutionHigh20sigma | DataProcessingResolutionHighOuterShell | DataProcessingRmergeOverall | DataProcessingRmergeLow | DataProcessingRmergeHigh | DataProcessingIsigOverall | DataProcessingIsigLow | DataProcessingIsigHigh | DataProcessingCompletenessOverall | DataProcessingCompletenessLow | DataProcessingCompletenessHigh | DataProcessingMultiplicityOverall | DataProcessingMultiplicityLow | DataProcessingMultiplicityHigh | DataProcessingCChalfOverall | DataProcessingCChalfLow | DataProcessingCChalfHigh | DataProcessingPathToLogfile | DataProcessingPathToMTZfile | DataProcessingLOGfileName | DataProcessingMTZfileName | DataProcessingDirectoryOriginal | DataProcessingUniqueReflectionsOverall | DataProcessingUniqueReflectionsLow | DataProcessingUniqueReflectionsHigh | DataProcessingLattice | DataProcessingPointGroup | DataProcessingUnitCellVolume | DataProcessingAlert | DataProcessingScore | DataProcessingStatus | DataProcessingRcryst | DataProcessingRfree | DataProcessingPathToDimplePDBfile | DataProcessingPathToDimpleMTZfile | DataProcessingDimpleSuccessful | DimpleResolutionHigh | DimpleRcryst | DimpleRfree | DimplePathToPDB | DimplePathToMTZ | DimpleReferencePDB | DimpleStatus | DimpleTwinResolutionHigh | DimpleTwinRcryst | DimpleTwinRfree | DimpleTwinPathToPDB | DimpleTwinPathToMTZ | DimpleTwinReferencePDB | DimpleTwinStatus | DimpleTwinFraction | DataProcessingDimpleTwinSuccessful | DimplePANDDAwasRun | DimplePANDDAhit | DimplePANDDAreject | DimplePANDDApath | PANDDAStatus | DatePanDDAModelCreated | RefinementResolution | RefinementResolutionTL | RefinementRcryst | RefinementRcrystTraficLight | RefinementRfree | RefinementRfreeTraficLight | RefinementSpaceGroup | RefinementLigandCC | RefinementRmsdBonds | RefinementRmsdBondsTL | RefinementRmsdAngles | RefinementRmsdAnglesTL | RefinementOutcome | RefinementOutcomePerson | RefinementOutcomeDate | RefinementMTZfree | RefinementTwinMTZfree | RefinementCIF | RefinementCIFStatus | RefinementCIFprogram | RefinementPDB_latest | RefinementMTZ_latest | RefinementMMCIFmodel_latest | RefinementMMCIFreflections_latest | RefinementMatrixWeight | RefinementComment | RefinementPathToRefinementFolder | RefinementLigandConfidence | RefinementLigandBoundConformation | RefinementBoundConformation | RefinementMolProbityScore | RefinementMolProbityScoreTL | RefinementRamachandranOutliers | RefinementRamachandranOutliersTL | RefinementRamachandranFavored | RefinementRamachandranFavoredTL | RefinementProgram | RefinementStatus | RefinementBusterReportHTML | RefinementRefiner | RefinementDate | Deposition_PDB_ID | Deposition_PDB_file | Deposition_Date | Deposition_mmCIF_model_file | Deposition_mmCIF_SF_file | Label | table_one | AssayIC50 | LastUpdated | LastUpdated_by
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
2039 | lb32633-6 | DSIPoised | F23 | DSIPoised | c1csc(n1)c2nnc[nH]2 | Z1041785508 | 9e5r_2024-04-29_RI1000-0276-3drop | A02d | -466 | 199 | 150 | CHIKV_MacB | 61 | 500 | 38.5 | 8 | 12.5 | done | 30/04/2024 12:11:10 |   |   |   |   |   |   | 01:48:50 | done | CHIKV_MacB-x1669 | ASAP-25 | 5 | DF075C1898 | OK: No comment:No comment | 30/04/2024 14:00:00 | 30/04/2024 14:00:00 | 0.0004328125 | exported to ispyb\lb32633-6_2024-04-30_14-33-18_ispyb | lb32633-41 |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   | i04-1 | 2024-05-02 02:03:23 | success | CHIKV_MacB-x1669_1 | cd79ecd2-ae46-4994-ac5c-c26d4962cd00 |   | 0.921237 | DF075C1898 | None | None | None | None | None | xia2-dials | P31 | 87.43437041 87.43437041 85.47407500 90.00000000 90.00000000 120.00000000 | True | 87.43437041 | 87.43437041 | 85.47407500 | 90.00000000 | 90.00000000 | 120.00000000 | 38.93 - 1.51 | 38.93 | 4.1 | 1.51 | n/a | None | 1.54 | 0.093 | 0.047 | 2.705 | 10.9 | 59.4 | 0.3 | 100.0 | 100.0 | 99.9 | 10.2 | 10.5 | 8.0 | 1.0 | 1.0 | 0.32 | /dls/labxchem/data/lb32633/lb32633-6/processing/analysis/model_building/CHIKV_MacB-x1669/autoprocessing/lb32633-41-CHIKV_MacB-x1669_1xia2-dials_cd79ecd2-ae46-4994-ac5c-c26d4962cd00/lb32633v41_xCHIKVMacBx16691_SAD_merging-statistics.json | /dls/labxchem/data/lb32633/lb32633-6/processing/analysis/model_building/CHIKV_MacB-x1669/autoprocessing/lb32633-41-CHIKV_MacB-x1669_1xia2-dials_cd79ecd2-ae46-4994-ac5c-c26d4962cd00/lb32633v41_xCHIKVMacBx16691_free.mtz | None | None | /dls/i04-1/data/2024/lb32633-41/processed/auto/CHIKV_MacB/CHIKV_MacB-x1669/CHIKV_MacB-x1669_1/cd79ecd2-ae46-4994-ac5c-c26d4962cd00/xia2-dials | 1176301 | 5750 | 5704 | hexagonal | 3 | 565886.6 | #00FF00 | 6797.30580296 | None |   |   |   |   | True |   | 0.19585 | 0.22804 | /dls/labxchem/data/lb32633/lb32633-6/processing/analysis/model_building/CHIKV_MacB-x1669/dimple.pdb | /dls/labxchem/data/lb32633/lb32633-6/processing/analysis/model_building/CHIKV_MacB-x1669/dimple.mtz | /dls/labxchem/data/lb32633/lb32633-6/processing/reference/CHIKV_MacB-x0257-ground-state.pdb | finished |   |   |   |   |   |   |   |   |   | True | True | False | /dls/labxchem/data/lb32633/lb32633-6/processing/analysis/panddas/20240502_pandda2_triplesoaks |   | 2024-05-03 13:45:35 | 1.51 | green | 0.19157 | green | 0.22817 | green | P 31 |   | 0.008 | green | 1.511 | green | 4 - CompChem ready | ill13029 | 2024-05-30_16-06-33.04 | CHIKV_MacB-x1669.free.mtz |   | /dls/labxchem/data/lb32633/lb32633-6/processing/analysis/model_building/CHIKV_MacB-x1669/compound/Z1041785508.cif | restraints generated | grade | /dls/science/groups/i04-1/AViDD/lb32633/lb32633-6/processing/analysis/model_building/CHIKV_MacB-x1669/Refine_0008/refine_8.pdb | /dls/science/groups/i04-1/AViDD/lb32633/lb32633-6/processing/analysis/model_building/CHIKV_MacB-x1669/Refine_0008/refine_8.mtz |   |   | 0.10239891 |   |   |   |   | /dls/science/groups/i04-1/AViDD/lb32633/lb32633-6/processing/analysis/model_building/CHIKV_MacB-x1669/Refine_0008/refine_8.split.bound-state.pdb | 1.24 | green | 0.00 | green | 99.53 | green |   | finished |   | dcn76311 | 2024-05-27_16-24-53.47 |   |   |   |   |   |   |   |   | 2024-05-30 16:06 | ill13029
```

I doubt it would be safe just to look for `*.cif` files as there is a lot of debris around. For instance, there's a file called `merged.cif` in that directory.

**Comment by mwinokan** [comment]
@tdudgeon. Thanks, I'll get back to you

**Comment by mwinokan** [comment]
@tdudgeon can you check if the RefinementCif file contains the aliases in the header like in the below under `desc_level`?

from Jasmin:

e.g. in x1689 (from the s drive where I have access to the cifs)
They are all listed in the cif quite up on the top:

# ---   LIST OF MONOMERS ---
#
data_comp_list
loop_
_chem_comp.id
_chem_comp.three_letter_code
_chem_comp.name
_chem_comp.group
_chem_comp.number_atoms_all
_chem_comp.number_atoms_nh
_chem_comp.desc_level
113      113 'unknown                             ' .                  11   7 .
756      756 'unknown                             ' .                  21  12 .
LIG      LIG 'unknown                             ' .                  12   6 .
# ------------------------------------------------------

**Comment by tdudgeon** [comment]
@mwinokan no it doesn't. For `x1669` the file that is referenced in soakdb (`Z1041785508.cif`) looks like this:

```
data_comp_list
#
loop_
_chem_comp.id 
_chem_comp.three_letter_code 
_chem_comp.name 
_chem_comp.group 
_chem_comp.number_atoms_all 
_chem_comp.number_atoms_nh 
_chem_comp.desc_level 
LIG LIG unknown . 14 10 . 
data_comp_LIG
loop_
_chem_comp_atom.comp_id 
_chem_comp_atom.atom_id 
_chem_comp_atom.type_symbol 
_chem_comp_atom.type_energy 
_chem_comp_atom.partial_charge 
_chem_comp_atom.x 
_chem_comp_atom.y 
_chem_comp_atom.z
```

The other 2 CIF's (not referenced in soakdb) also just mention themselves.

**Comment by mwinokan** [comment]
OK. I'll liaise with Daren/Jasmin. We might have to come up with a new SoP for the diamond processing

**Comment by mwinokan** [comment]
@tdudgeon could you please contact Daren and hash out with him how we need to change the SoakDB SoP for the combi-soak CIF's to be in the correct place and format for XCA and Fragalysis.

**Comment by tdudgeon** [comment]
I spoke with Daren about this and he says that the way that this needs to be handled is for the user to _manually_ merge the multiple CIF files into a single file, keeping the separate residue IDs and ensuring these are consistent with the equivalent residue IDs in the PDB file. The `RefinementCIF` record in soadkDB will need to be updated to point to that merged file.

Once that is done, XCA will need to be updated to handle this (including a change to the "model" to allow multiple compounds to be described), then the target loader needs to be changed. The Fragalysis model and API may need some changes, but @kaliif says the model already handles multiple compounds per crystal. The front end will also likely need some changes as well/

**Comment by tdudgeon** [comment]
I worked with Jasmin to get an example for this, and then discussed the mechanics of what needs to happen with @ConorFWild.
The changes to _collator_ turned out to be relatively straightforward, and the results now look like this:

```
    crystallographic_files:
      xtal_pdb:
        file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739.pdb
        sha256: 409a78c98319676418f5c93dab0e177fd0d18b5981f74c61829eb29a2aba03c3
      xtal_mtz:
        file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739.mtz
        sha256: ab258ed61adfd11e4312fcfbb7218d50bcb5604290feaa610f8f741248a20146
      ligand_cif:
        file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739.cif
        sha256: d0556ef2b0c3bd065be7d29a3b37698bd1cb099a5122cb27f6fa6a75231a2a98
        ligands:
          LIG:
            smiles: O=C1NCCN1
          LG1:
            smiles: Cc1cn[nH]c1-c1ccncc1
          LG2:
            smiles: Nc1nnc2ccccn12
      ligand_binding_events:
      - file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739_1_A_401.ccp4
        sha256: 4ab0da714da26bcd079252f6da792020d83dafeddab8ca130391d171694f9de6
        model: '1'
        chain: A
        res: 401
        index: 1
        bdc: 0.24
      - file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739_1_A_402.ccp4
        sha256: 4ab0da714da26bcd079252f6da792020d83dafeddab8ca130391d171694f9de6
        model: '1'
        chain: A
        res: 402
        index: 1
        bdc: 0.24
      - file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739_1_A_403.ccp4
        sha256: 4ab0da714da26bcd079252f6da792020d83dafeddab8ca130391d171694f9de6
        model: '1'
        chain: A
        res: 403
        index: 1
        bdc: 0.24
      - file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739_1_B_401.ccp4
        sha256: 4ab0da714da26bcd079252f6da792020d83dafeddab8ca130391d171694f9de6
        model: '1'
        chain: B
        res: 401
        index: 1
        bdc: 0.24
      - file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739_1_B_402.ccp4
        sha256: 4ab0da714da26bcd079252f6da792020d83dafeddab8ca130391d171694f9de6
        model: '1'
        chain: B
        res: 402
        index: 1
        bdc: 0.24
      - file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739_1_B_403.ccp4
        sha256: 4ab0da714da26bcd079252f6da792020d83dafeddab8ca130391d171694f9de6
        model: '1'
        chain: B
        res: 403
        index: 1
        bdc: 0.24
      - file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739_1_C_401.ccp4
        sha256: def2026f8d1c18a7646c2acfbdf636211695e7f1c99def86c60dee3daebc52ff
        model: '1'
        chain: C
        res: 401
        index: 2
        bdc: 0.31
      - file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739_1_C_402.ccp4
        sha256: def2026f8d1c18a7646c2acfbdf636211695e7f1c99def86c60dee3daebc52ff
        model: '1'
        chain: C
        res: 402
        index: 2
        bdc: 0.31
      - file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739_1_C_403.ccp4
        sha256: def2026f8d1c18a7646c2acfbdf636211695e7f1c99def86c60dee3daebc52ff
        model: '1'
        chain: C
        res: 403
        index: 2
        bdc: 0.31
      - file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739_1_D_401.ccp4
        sha256: 99c2b9c507689e12a939223e3196322d2956c741c0754fc6f97762836e0305fe
        model: '1'
        chain: D
        res: 401
        index: 3
        bdc: 0.23
      - file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739_1_D_402.ccp4
        sha256: 99c2b9c507689e12a939223e3196322d2956c741c0754fc6f97762836e0305fe
        model: '1'
        chain: D
        res: 402
        index: 3
        bdc: 0.23
      - file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739_1_D_403.ccp4
        sha256: 99c2b9c507689e12a939223e3196322d2956c741c0754fc6f97762836e0305fe
        model: '1'
        chain: D
        res: 403
        index: 3
        bdc: 0.23
```

The differences from before are:

1. the `ligand_cif` section now contains a `ligands` section which contains a dict of the ligand data. Previously it was assumed that there was only a single ligand for each crystal
2. the `ligand_binding_events` section now contains additional events for the additional ligands.

In terms of how the inputs must be generated:

1. the multiple ligands must be merged into a single CIF file, and referenced from soakdb
2. the ligands must be named LIG, LG1, LG2, LG3, ...
3. the names of the ligands in the PDB must be consistent with those in the CIF file

According to Jasmin the only way to achieve this at present is by manual manipulation of the XCE data.
These requirements will need to be documented and communicated with the users.

**Note** that the current code is not optimal. The ligand binding events are determined entirely from the data in the PDB file, while the data in the `ligands` section comes entirely from the CIF file. There is currently the assumption that these two files are completely consistent. We should add some checks to ensure this.

**Note** that the _aligner_ tool still needs to be worked on before complete data can be generated.

**Note** that the event map `.ccp4` files look like they are somewhat duplicated (as seen by their checksums), and as these are large this should be avoided. This is not a new feature, just more obvious now that we have more of them.

This work is currently on the `combi_soaks` branch.

**Comment by mwinokan** [comment]
@tdudgeon what is the technical reason for requiring the strict ligand naming: LIG, LG1, LG2, LG3, ... ? The user may wish to use standardised ligand names. And is it limited to three characters?

**Comment by tdudgeon** [comment]
Technically speaking there is no absolute need for this, but:
1. it makes it straightforward to identify the ligands of interest (as opposed to cofactors, salts, solvents etc.). If we can't use the names then we'd need to find another mechanism.
2. they have to be named something (that has 3 letters) so why not give them sensible and consistent names.

**Comment by mwinokan** [comment]
@tdudgeon there is some example combi-soak data [on the legacy stack](https://fragalysis-legacy.xchem.diamond.ac.uk/viewer/react/projects/4020/3563). Is there any way to bodge that for testing, or would it need a full alignment with XCA?

**Comment by tdudgeon** [comment]
@mwinokan no, the whole point is that it has to be run through the XCA process.

**Comment by tdudgeon** [comment]
As I'm not that familiar with CIF format or the [gemmi Python library](https://project-gemmi.github.io/python-api/gemmi.cif.html), I'd like people (mostly @ConorFWild) to do a sanity check on what I am proposing for handling the case when there are multiple ligands in a structure.

1. A single CIF file (referenced from soakDB) is used as the statement of truth when it comes to defining the ligands of interest (e.g. users may need to do some manual work to create this file)
2. We first look at this CIF file first as if it was as is seems to be defined at Diamond when there are multiple ligands present:
   - read the block called `data-comp_list` which defined the ligand blocks of interest in the remainder of the file
   - if that block exists read the ligand IDs from this block
   - read the block corresponding to each ligand ID to get the actual data for each ligand of interest
3. If it does not contain the block named `data_comp_list` then we assume it is either a CIF from the PDB or a CIF from Diamond with only a single ligand, and in either case there will be a SINGLE block present, so we use that block.

**Note**: this removes the need to name the ligands LIG, LG12, LG2 ..., but I still recommend we do this for clarity.

To make this clearer below is an example multi-ligand CIF file and some simple Python code that exemplifies this.

### CIF file

```
global_
_lib_name         ?
_lib_version      ?
_lib_update       ?
# ------------------------------------------------
#
# ---   LIST OF MONOMERS ---
#
data_comp_list
loop_
_chem_comp.id
_chem_comp.three_letter_code
_chem_comp.name
_chem_comp.group
_chem_comp.number_atoms_all
_chem_comp.number_atoms_nh
_chem_comp.desc_level
113      113 'unknown                             ' .                  11   7 .
756      756 'unknown                             ' .                  21  12 .
LIG      LIG 'unknown                             ' .                  12   6 .
# ------------------------------------------------------
#
# --- DESCRIPTION OF MONOMERS ---
#
data_comp_113
#
loop_
_chem_comp_atom.comp_id
_chem_comp_atom.atom_id
_chem_comp_atom.type_symbol
_chem_comp_atom.type_energy
_chem_comp_atom.partial_charge
_chem_comp_atom.x
_chem_comp_atom.y
_chem_comp_atom.z
 113           O      O    O         0.000      4.506    0.400    2.414
 113           C      C    CR6       0.000      4.646   -0.652    1.773
 113           N1     N    NR16      0.000      4.278   -1.840    2.323
 113           H3     H    HNR6      0.000      3.886   -1.893    3.257
 113           N      N    NRD6      0.000      4.371   -3.070    1.741
 113           C3     C    CR16      0.000      4.860   -3.109    0.546
 113           H2     H    HCR       0.000      4.922   -4.105    0.109
 113           C2     C    CR16      0.000      5.287   -1.974   -0.162
 113           H1     H    HCR       0.000      5.690   -2.077   -1.169
 113           C1     C    CR16      0.000      5.181   -0.757    0.445
 113           H      H    HCR       0.000      5.492    0.165   -0.043
loop_
_chem_comp_tree.comp_id
_chem_comp_tree.atom_id
_chem_comp_tree.atom_back
_chem_comp_tree.atom_forward
_chem_comp_tree.connect_type
 113      O      n/a    C      START
 113      C      O      N1     .
 113      N1     C      N      .
 113      H3     N1     .      .
 113      N      N1     C3     .
 113      C3     N      C2     .
 113      H2     C3     .      .
 113      C2     C3     C1     .
 113      H1     C2     .      .
 113      C1     C2     H      .
 113      H      C1     .      END
 113      C      C1     .    ADD
loop_
_chem_comp_bond.comp_id
_chem_comp_bond.atom_id_1
_chem_comp_bond.atom_id_2
_chem_comp_bond.type
_chem_comp_bond.value_dist
_chem_comp_bond.value_dist_esd
 113      C      O         double      1.240    0.016
 113      C      C1        single      1.436    0.015
 113      C1     C2        double      1.363    0.017
 113      C2     C3        single      1.402    0.015
 113      C3     N         double      1.291    0.010
 113      N      N1        single      1.365    0.013
 113      N1     C         single      1.361    0.014
 113      H      C1        single      1.089    0.020
 113      H1     C2        single      1.089    0.020
 113      H2     C3        single      1.089    0.020
 113      H3     N1        single      1.015    0.020
loop_
_chem_comp_angle.comp_id
_chem_comp_angle.atom_id_1
_chem_comp_angle.atom_id_2
_chem_comp_angle.atom_id_3
_chem_comp_angle.value_angle
_chem_comp_angle.value_angle_esd
 113      O      C      N1      120.000    0.700
 113      O      C      C1      125.100    1.300
 113      N1     C      C1      114.400    1.100
 113      C      N1     H3      121.500    3.000
 113      C      N1     N       126.900    0.800
 113      H3     N1     N       111.800    3.000
 113      N1     N      C3      116.700    0.800
 113      N      C3     H2      114.900    3.000
 113      N      C3     C2      123.800    3.000
 113      H2     C3     C2      121.400    3.000
 113      C3     C2     H1      120.100    3.000
 113      C3     C2     C1      118.300    3.000
 113      H1     C2     C1      121.600    3.000
 113      C2     C1     H       122.200    3.000
 113      C2     C1     C       120.600    1.000
 113      H      C1     C       117.400    3.000
loop_
_chem_comp_tor.comp_id
_chem_comp_tor.id
_chem_comp_tor.atom_id_1
_chem_comp_tor.atom_id_2
_chem_comp_tor.atom_id_3
_chem_comp_tor.atom_id_4
_chem_comp_tor.value_angle
_chem_comp_tor.value_angle_esd
_chem_comp_tor.period
 113      CONST_ring6A-6   C1     C      N1     N                  0.000            0.000   0
 113      CONST_ring6A-5   N1     C      C1     C2                 0.000            0.000   0
 113      CONST_ring6A-1   C      N1     N      C3                 0.000            0.000   0
 113      CONST_ring6A-2   N1     N      C3     C2                 0.000            0.000   0
 113      CONST_ring6A-3   N      C3     C2     C1                 0.000            0.000   0
 113      CONST_ring6A-4   C3     C2     C1     C                  0.000            0.000   0
loop_
_chem_comp_plane_atom.comp_id
_chem_comp_plane_atom.plane_id
_chem_comp_plane_atom.atom_id
_chem_comp_plane_atom.dist_esd
 113      csd-C             C         0.020
 113      csd-C             O         0.020
 113      csd-C             N1        0.020
 113      csd-C             C1        0.020
 113      qm-C1             C1        0.020
 113      qm-C1             C         0.020
 113      qm-C1             C2        0.020
 113      qm-C1             H         0.020
 113      qm-C2             C2        0.020
 113      qm-C2             C3        0.020
 113      qm-C2             C1        0.020
 113      qm-C2             H1        0.020
 113      qm-C3             C3        0.020
 113      qm-C3             N         0.020
 113      qm-C3             C2        0.020
 113      qm-C3             H2        0.020
 113      qm-N1             N1        0.020
 113      qm-N1             C         0.020
 113      qm-N1             N         0.020
 113      qm-N1             H3        0.020
 113      ring6A-1          C         0.020
 113      ring6A-1          N1        0.020
 113      ring6A-1          N         0.020
 113      ring6A-1          C3        0.020
 113      ring6A-2          N1        0.020
 113      ring6A-2          N         0.020
 113      ring6A-2          C3        0.020
 113      ring6A-2          C2        0.020
 113      ring6A-3          N         0.020
 113      ring6A-3          C3        0.020
 113      ring6A-3          C2        0.020
 113      ring6A-3          C1        0.020
 113      ring6A-4          C3        0.020
 113      ring6A-4          C2        0.020
 113      ring6A-4          C1        0.020
 113      ring6A-4          C         0.020
 113      ring6A-5          C2        0.020
 113      ring6A-5          C1        0.020
 113      ring6A-5          C         0.020
 113      ring6A-5          N1        0.020
 113      ring6A-6          C1        0.020
 113      ring6A-6          C         0.020
 113      ring6A-6          N1        0.020
 113      ring6A-6          N         0.020
#
data_comp_756
#
loop_
_chem_comp_atom.comp_id
_chem_comp_atom.atom_id
_chem_comp_atom.type_symbol
_chem_comp_atom.type_energy
_chem_comp_atom.partial_charge
_chem_comp_atom.x
_chem_comp_atom.y
_chem_comp_atom.z
 756           C      C    CH3       0.000      3.382    3.584    2.273
 756           H1     H    HCH3      0.000      3.834    3.973    1.362
 756           H2     H    HCH3      0.000      4.054    2.820    2.661
 756           H3     H    HCH3      0.000      3.338    4.402    2.990
 756           C1     C    CR5       0.000      2.006    3.044    2.003
 756           C3     C    CR5       0.000      0.847    3.749    1.682
 756           N1     N    NR15      0.000     -0.130    2.826    1.534
 756           H      H    HNR5      0.000     -1.107    2.969    1.300
 756           N      N    NRD5      0.000      0.316    1.556    1.738
 756           C2     C    CR15      0.000      1.608    1.697    2.021
 756           H4     H    HCR       0.000      2.245    0.840    2.235
 756           C4     C    CR6       0.000      0.586    5.191    1.507
 756           C8     C    CR16      0.000      1.019    6.113    2.448
 756           H8     H    HCR       0.000      1.563    5.829    3.348
 756           C7     C    CR16      0.000      0.752    7.452    2.240
 756           H7     H    HCR       0.000      1.070    8.214    2.949
 756           N2     N    NRD6      0.000      0.094    7.920    1.175
 756           C6     C    CR16      0.000     -0.317    7.020    0.276
 756           H6     H    HCR       0.000     -0.849    7.434   -0.579
 756           C5     C    CR16      0.000     -0.098    5.662    0.396
 756           H5     H    HCR       0.000     -0.461    4.984   -0.376
loop_
_chem_comp_tree.comp_id
_chem_comp_tree.atom_id
_chem_comp_tree.atom_back
_chem_comp_tree.atom_forward
_chem_comp_tree.connect_type
 756      C      n/a    C1     START
 756      H1     C      .      .
 756      H2     C      .      .
 756      H3     C      .      .
 756      C1     C      C3     .
 756      C3     C1     C4     .
 756      N1     C3     N      .
 756      H      N1     .      .
 756      N      N1     C2     .
 756      C2     N      H4     .
 756      H4     C2     .      .
 756      C4     C3     C8     .
 756      C8     C4     C7     .
 756      H8     C8     .      .
 756      C7     C8     N2     .
 756      H7     C7     .      .
 756      N2     C7     C6     .
 756      C6     N2     C5     .
 756      H6     C6     .      .
 756      C5     C6     H5     .
 756      H5     C5     .      END
 756      C1     C2     .    ADD
 756      C4     C5     .    ADD
loop_
_chem_comp_bond.comp_id
_chem_comp_bond.atom_id_1
_chem_comp_bond.atom_id_2
_chem_comp_bond.type
_chem_comp_bond.value_dist
_chem_comp_bond.value_dist_esd
 756      C1     C         single      1.502    0.018
 756      C1     C2        aromatic    1.402    0.014
 756      C2     N         aromatic    1.329    0.010
 756      N      N1        aromatic    1.365    0.015
 756      H      N1        single      1.015    0.020
 756      N1     C3        aromatic    1.353    0.009
 756      C3     C1        aromatic    1.391    0.022
 756      C4     C3        aromatic    1.476    0.011
 756      C4     C5        aromatic    1.391    0.014
 756      C5     C6        aromatic    1.381    0.011
 756      C6     N2        aromatic    1.334    0.011
 756      N2     C7        aromatic    1.334    0.011
 756      C7     C8        aromatic    1.381    0.011
 756      C8     C4        aromatic    1.391    0.014
 756      H1     C         single      1.089    0.020
 756      H2     C         single      1.089    0.020
 756      H3     C         single      1.089    0.020
 756      H4     C2        single      1.089    0.020
 756      H5     C5        single      1.089    0.020
 756      H6     C6        single      1.089    0.020
 756      H7     C7        single      1.089    0.020
 756      H8     C8        single      1.089    0.020
loop_
_chem_comp_angle.comp_id
_chem_comp_angle.atom_id_1
_chem_comp_angle.atom_id_2
_chem_comp_angle.atom_id_3
_chem_comp_angle.value_angle
_chem_comp_angle.value_angle_esd
 756      H1     C      H2      107.000    3.000
 756      H1     C      H3      107.400    3.000
 756      H2     C      H3      108.500    3.000
 756      H1     C      C1      111.000    3.000
 756      H2     C      C1      112.200    3.000
 756      H3     C      C1      110.600    3.000
 756      C      C1     C3      128.400    1.100
 756      C      C1     C2      126.900    3.000
 756      C3     C1     C2      104.500    0.700
 756      C1     C3     N1      106.300    0.600
 756      C1     C3     C4      131.500    1.500
 756      N1     C3     C4      121.500    1.300
 756      C3     N1     H       128.600    3.000
 756      C3     N1     N       112.700    0.500
 756      H      N1     N       118.800    3.000
 756      N1     N      C2      105.200    1.600
 756      N      C2     H4      121.800    3.000
 756      N      C2     C1      111.900    0.900
 756      H4     C2     C1      126.300    3.000
 756      C3     C4     C8      120.700    0.900
 756      C3     C4     C5      120.700    0.900
 756      C8     C4     C5      118.700    0.900
 756      C4     C8     H8      122.900    3.000
 756      C4     C8     C7      119.000    0.600
 756      H8     C8     C7      118.400    3.000
 756      C8     C7     H7      121.500    3.000
 756      C8     C7     N2      123.800    0.900
 756      H7     C7     N2      114.700    3.000
 756      C7     N2     C6      116.600    1.100
 756      N2     C6     H6      115.000    3.000
 756      N2     C6     C5      123.800    0.900
 756      H6     C6     C5      121.300    3.000
 756      C6     C5     H5      119.900    3.000
 756      C6     C5     C4      119.000    0.600
 756      H5     C5     C4      121.400    3.000
loop_
_chem_comp_tor.comp_id
_chem_comp_tor.id
_chem_comp_tor.atom_id_1
_chem_comp_tor.atom_id_2
_chem_comp_tor.atom_id_3
_chem_comp_tor.atom_id_4
_chem_comp_tor.value_angle
_chem_comp_tor.value_angle_esd
_chem_comp_tor.period
 756      other-001        H1     C      C1     C3                 0.000      1000000.000  10
 756      CONST_ring5A-5   C2     C1     C3     N1                 0.000            0.000   0
 756      CONST_ring5A-4   C3     C1     C2     N                  0.000            0.000   0
 756      CONST_ring5A-1   C1     C3     N1     N                  0.000            0.000   0
 756      qmf-01           C1     C3     C4     C8               180.000      1000000.000   2
 756      CONST_ring5A-2   C3     N1     N      C2                 0.000            0.000   0
 756      CONST_ring5A-3   N1     N      C2     C1                 0.000            0.000   0
 756      CONST_ring6A-6   C5     C4     C8     C7                 0.000            0.000   0
 756      CONST_ring6A-5   C8     C4     C5     C6                 0.000            0.000   0
 756      CONST_ring6A-1   C4     C8     C7     N2                 0.000            0.000   0
 756      CONST_ring6A-2   C8     C7     N2     C6                 0.000            0.000   0
 756      CONST_ring6A-3   C7     N2     C6     C5                 0.000            0.000   0
 756      CONST_ring6A-4   N2     C6     C5     C4                 0.000            0.000   0
loop_
_chem_comp_plane_atom.comp_id
_chem_comp_plane_atom.plane_id
_chem_comp_plane_atom.atom_id
_chem_comp_plane_atom.dist_esd
 756      csd-C3            C3        0.020
 756      csd-C3            C1        0.020
 756      csd-C3            N1        0.020
 756      csd-C3            C4        0.020
 756      csd-C4            C4        0.020
 756      csd-C4            C3        0.020
 756      csd-C4            C8        0.020
 756      csd-C4            C5        0.020
 756      qm-C1             C1        0.020
 756      qm-C1             C         0.020
 756      qm-C1             C3        0.020
 756      qm-C1             C2        0.020
 756      qm-C2             C2        0.020
 756      qm-C2             C1        0.020
 756      qm-C2             N         0.020
 756      qm-C2             H4        0.020
 756      qm-C5             C5        0.020
 756      qm-C5             C4        0.020
 756      qm-C5             C6        0.020
 756      qm-C5             H5        0.020
 756      qm-C6             C6        0.020
 756      qm-C6             N2        0.020
 756      qm-C6             C5        0.020
 756      qm-C6             H6        0.020
 756      qm-C7             C7        0.020
 756      qm-C7             C8        0.020
 756      qm-C7             N2        0.020
 756      qm-C7             H7        0.020
 756      qm-C8             C8        0.020
 756      qm-C8             C4        0.020
 756      qm-C8             C7        0.020
 756      qm-C8             H8        0.020
 756      qm-N1             N1        0.020
 756      qm-N1             C3        0.020
 756      qm-N1             N         0.020
 756      qm-N1             H         0.020
 756      ring5A-1          C1        0.020
 756      ring5A-1          C3        0.020
 756      ring5A-1          N1        0.020
 756      ring5A-1          N         0.020
 756      ring5A-2          C3        0.020
 756      ring5A-2          N1        0.020
 756      ring5A-2          N         0.020
 756      ring5A-2          C2        0.020
 756      ring5A-3          N1        0.020
 756      ring5A-3          N         0.020
 756      ring5A-3          C2        0.020
 756      ring5A-3          C1        0.020
 756      ring5A-4          N         0.020
 756      ring5A-4          C2        0.020
 756      ring5A-4          C1        0.020
 756      ring5A-4          C3        0.020
 756      ring5A-5          C2        0.020
 756      ring5A-5          C1        0.020
 756      ring5A-5          C3        0.020
 756      ring5A-5          N1        0.020
 756      ring6A-1          C4        0.020
 756      ring6A-1          C8        0.020
 756      ring6A-1          C7        0.020
 756      ring6A-1          N2        0.020
 756      ring6A-2          C8        0.020
 756      ring6A-2          C7        0.020
 756      ring6A-2          N2        0.020
 756      ring6A-2          C6        0.020
 756      ring6A-3          C7        0.020
 756      ring6A-3          N2        0.020
 756      ring6A-3          C6        0.020
 756      ring6A-3          C5        0.020
 756      ring6A-4          N2        0.020
 756      ring6A-4          C6        0.020
 756      ring6A-4          C5        0.020
 756      ring6A-4          C4        0.020
 756      ring6A-5          C6        0.020
 756      ring6A-5          C5        0.020
 756      ring6A-5          C4        0.020
 756      ring6A-5          C8        0.020
 756      ring6A-6          C5        0.020
 756      ring6A-6          C4        0.020
 756      ring6A-6          C8        0.020
 756      ring6A-6          C7        0.020
#
data_comp_LIG
#
loop_
_chem_comp_atom.comp_id
_chem_comp_atom.atom_id
_chem_comp_atom.type_symbol
_chem_comp_atom.type_energy
_chem_comp_atom.partial_charge
_chem_comp_atom.x
_chem_comp_atom.y
_chem_comp_atom.z
 LIG           O      O    O         0.000      1.393   -0.681    4.217
 LIG           C      C    C         0.000      1.496   -0.871    3.007
 LIG           N1     N    NH1       0.000      2.557   -1.405    2.394
 LIG           H5     H    H         0.000      3.402   -1.715    2.863
 LIG           C2     C    CH2       0.000      2.378   -1.502    0.951
 LIG           H3     H    H         0.000      2.463   -2.528    0.595
 LIG           H4     H    H         0.000      3.149   -0.952    0.413
 LIG           C1     C    CH2       0.000      0.982   -0.922    0.740
 LIG           H1     H    H         0.000      0.296   -1.630    0.277
 LIG           H2     H    H         0.000      0.984   -0.056    0.079
 LIG           N      N    NH1       0.000      0.565   -0.578    2.093
 LIG           H      H    H         0.000     -0.338   -0.162    2.299
loop_
_chem_comp_tree.comp_id
_chem_comp_tree.atom_id
_chem_comp_tree.atom_back
_chem_comp_tree.atom_forward
_chem_comp_tree.connect_type
 LIG      O      n/a    C      START
 LIG      C      O      N      .
 LIG      N1     C2     H5     .
 LIG      H5     N1     .      END
 LIG      C2     C1     N1     .
 LIG      H3     C2     .      .
 LIG      H4     C2     .      .
 LIG      C1     N      C2     .
 LIG      H1     C1     .      .
 LIG      H2     C1     .      .
 LIG      N      C      C1     .
 LIG      H      N      .      .
 LIG      C      N1     .    ADD
loop_
_chem_comp_bond.comp_id
_chem_comp_bond.atom_id_1
_chem_comp_bond.atom_id_2
_chem_comp_bond.type
_chem_comp_bond.value_dist
_chem_comp_bond.value_dist_esd
 LIG      C      O         double      1.229    0.013
 LIG      C      N         single      1.338    0.014
 LIG      N      C1        single      1.457    0.013
 LIG      C1     C2        single      1.525    0.016
 LIG      C2     N1        single      1.457    0.013
 LIG      N1     C         single      1.338    0.014
 LIG      H      N         single      1.015    0.020
 LIG      H1     C1        single      1.089    0.020
 LIG      H2     C1        single      1.089    0.020
 LIG      H3     C2        single      1.089    0.020
 LIG      H4     C2        single      1.089    0.020
 LIG      H5     N1        single      1.015    0.020
loop_
_chem_comp_angle.comp_id
_chem_comp_angle.atom_id_1
_chem_comp_angle.atom_id_2
_chem_comp_angle.atom_id_3
_chem_comp_angle.value_angle
_chem_comp_angle.value_angle_esd
 LIG      O      C      N1      125.700    1.000
 LIG      O      C      N       125.700    1.000
 LIG      N1     C      N       109.000    1.200
 LIG      C      N1     H5      124.700    3.000
 LIG      C      N1     C2      112.100    1.200
 LIG      H5     N1     C2      122.600    3.000
 LIG      N1     C2     H3      112.200    3.000
 LIG      N1     C2     H4      111.700    3.000
 LIG      N1     C2     C1      102.400    1.100
 LIG      H3     C2     H4      105.100    3.000
 LIG      H3     C2     C1      112.700    3.000
 LIG      H4     C2     C1      112.900    3.000
 LIG      C2     C1     H1      112.900    3.000
 LIG      C2     C1     H2      112.700    3.000
 LIG      C2     C1     N       102.400    1.100
 LIG      H1     C1     H2      105.100    3.000
 LIG      H1     C1     N       111.700    3.000
 LIG      H2     C1     N       112.200    3.000
 LIG      C1     N      H       122.600    3.000
 LIG      C1     N      C       112.100    1.200
 LIG      H      N      C       124.700    3.000
loop_
_chem_comp_tor.comp_id
_chem_comp_tor.id
_chem_comp_tor.atom_id_1
_chem_comp_tor.atom_id_2
_chem_comp_tor.atom_id_3
_chem_comp_tor.atom_id_4
_chem_comp_tor.value_angle
_chem_comp_tor.value_angle_esd
_chem_comp_tor.period
 LIG      qmf-01           N      C      N1     C2               180.000      1000000.000   2
 LIG      qmf-02           N1     C      N      C1               180.000      1000000.000   2
 LIG      other-001        C      N1     C2     C1                 0.000      1000000.000  10
 LIG      other-002        N1     C2     C1     N                  0.000      1000000.000  10
 LIG      other-003        C2     C1     N      C                  0.000      1000000.000  10
loop_
_chem_comp_plane_atom.comp_id
_chem_comp_plane_atom.plane_id
_chem_comp_plane_atom.atom_id
_chem_comp_plane_atom.dist_esd
 LIG      csd-C             C         0.020
 LIG      csd-C             O         0.020
 LIG      csd-C             N1        0.020
 LIG      csd-C             N         0.020
 LIG      qm-N              N         0.020
 LIG      qm-N              C         0.020
 LIG      qm-N              C1        0.020
 LIG      qm-N              H         0.020
 LIG      qm-N1             N1        0.020
 LIG      qm-N1             C         0.020
 LIG      qm-N1             C2        0.020
 LIG      qm-N1             H5        0.020
 LIG      qmf-01            N         0.020
 LIG      qmf-01            C         0.020
 LIG      qmf-01            N1        0.020
 LIG      qmf-01            C2        0.020
 LIG      qmf-02            N1        0.020
 LIG      qmf-02            C         0.020
 LIG      qmf-02            N         0.020
 LIG      qmf-02            C1        0.020

# ------------------------------------------------------
# ------------------------------------------------------
#
# --- DESCRIPTION OF MONOMERS ---
#
# ------------------------------------------------------
```

### Python code
```
from gemmi import cif

cif_file = 'data/lb32633-6-combi/inputs/pdbs/x1689.cif'
# cif_file = 'data/lb32633-6-combi/inputs/pdbs/7h7f.cif'

doc = cif.read(cif_file)

ligand_blocks = []
list_block = doc.find_block('comp_list')
if list_block: # seems to be a multi-ligand CIF
    components = list_block.find_loop('_chem_comp.id')
    for component in components:
        block = doc.find_block('comp_' + str(component))
        print(component, block)
        ligand_blocks.append(block)
else: # seems to be a single ligand CIF
    print('must be single mol CIF e.g. from PDB')
    block = doc.sole_block()
    print(block)
    ligand_blocks.append(block)

print('Found', len(ligand_blocks), 'ligand blocks')
```

**Comment by phraenquex** [comment]
@tdudgeon this looks fine to me.  

Indeed, all that matters is that the behaviour is very clearly defined, and the instructions are well-written (i.e. test-driven by at least 2 non-expert users).

**Comment by tdudgeon** [comment]
What I forgot to mention was that I'm proposing that the contents of the CIF file define the ligands that the _aligner_ tool handles e.g. the CIF is the statement of truth for what is considered a ligand in Fragalysis.
I fear that @ConorFWild may be less comfortable with this.

**Comment by phraenquex** [comment]
I thought @ConorFWild would be _very_ comfortable with it.  We definitely need _some_ way for users to tell the aligner what is the ground truth - relying just on LIG,LG1,LG2... is bad juju.  

@ConorFWild please comment.

**Comment by tdudgeon** [comment]
Initial extraction of ligand info from CIF file is done.
Now teed to liaise with @ConorFWild on passing this info into the alignments.

**Comment by mwinokan** [comment]
Other combi-soak data is not prepared for v2 and to avoid the overhead of the XCA processing we will wait for #1477 to be implemented and use the CHIKV_Mac data

**Comment by mwinokan** [comment]
This ticket will need some bandwidth from @ConorFWild

**Comment by mwinokan** [comment]
@tdudgeon says we are now using the ligand CIF file as the ground truth. This has caused a change in the meta_collator.yaml:

Previously there was the assumption that there is only a single ligand. There can now be more, and there are key value pairs for ligand name and smiles.

@phraenquex asks if it necessary to place constraints on the residue names of the ligands. @tdudgeon says we can cope with arbitrary ligand names as long as they are in the CIF.

**Comment by tdudgeon** [comment]
Aligner ran without problems and it might even have generated all the right outputs.
It's complex to assess and probably needs multiple people to look at it. @ConorFWild , @mwinokan could you take a look?

This attached file is for just one crystal, and there are 4 chains and 3 ligands, so there's lots of output.
[x1739.tar.gz](https://github.com/user-attachments/files/17149858/x1739.tar.gz)

Important to point out that the inputs need to be correct. There's another crystal where the ligands are not correctly assigned in the PDB and the output for that one is a bit of a mess.

**Comment by mwinokan** [comment]
Thanks @tdudgeon, I will cautiously say that it seems to be a success! All 12 observations have been aligned:

![Screenshot 2024-09-26 at 15 13 55](https://github.com/user-attachments/assets/d1239d26-43be-4913-8f69-886f557bb575)

**Comment by tdudgeon** [comment]
To test this more thoroughly we'll need some better input data. The initial set I was given by Jasmin was a bit strange and some of the structures were not properly defined (ligands asssigned to the wrong chain etc.)

**Comment by mwinokan** [comment]
> ligands assigned to the wrong chain

@tdudgeon are there any other issues except the wrong chain assignments? Those will be fixed once Jasmin aligns CHIKV_Mac with the #1511 changes

**Comment by tdudgeon** [comment]
I think the ligand naming was inconsistent, but that wouldn't cause problems, just a bit of confusion.

**Comment by mwinokan** [comment]
@ConorFWild please check that XCA does not align any ligands that are found in the PDB file but are not present in the CIF

**Comment by tdudgeon** [comment]
A question for @ConorFWild . We encountered an issue with how the ligands are now processed. Recall that the YAML now looks like this:
```
      ligand_cif:
        file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739.cif
        sha256: d0556ef2b0c3bd065be7d29a3b37698bd1cb099a5122cb27f6fa6a75231a2a98
        source_file: data/lb32633-6/inputs/dls/labxchem/data/lb32633/lb32633-6/processing/analysis/model_building/CHIKV_MacB-x1739/merged.cif
        ligands:
          LG1:
            smiles: Cc1cn[nH]c1-c1ccncc1
          LG2:
            smiles: Nc1nnc2ccccn12
          LIG:
            smiles: O=C1NCCN1
      ligand_binding_events:
      - file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739_1_A_401.ccp4
        sha256: cc18731ba7a535d696180fab2f3e2bf98c3bde0c536768608cef85a2f526c0cd
        source_file: data/lb32633-6/inputs/dls/labxchem/data/lb32633/lb32633-6/processing/analysis/panddas/20240730_pandda2_1560-2471/processed_datasets/CHIKV_MacB-x1739/CHIKV_MacB-x1739-event_3_1-BDC_0.19_map.native.ccp4
        model: '1'
        chain: A
        res: 401
        index: 3
        bdc: 0.19
```
(only one `ligand_binding_events` entry is shown, but that's enough to illustrate the problem).

@kaliif thinks that the target loader needs to be able to correlate each `ligand_binding_events` entry with it's corresponding ligand in the `ligands` section. Initially I thought that could be done by adding the residue number to the `ligands` section, but that won't work as the residue numbers in different chains could potentially be different (and that info is not available in the CIF file). So instead we need to include the ligand name (e.g. `LIG`) in the `ligand_binding_events` section but the data that comes back from LNA does not include this. Presumably it is available in LNA and could be added to the dictionary. So is it possible to do this?

**Comment by phraenquex** [comment]
@ConorFWild to summarise:  the `ligand_binding_event` field needs an additional field, `resname` that points to the relevant `ligands` in `ligand_cif`. 

@tdudgeon could you have a try at finding where this might be?

**Comment by tdudgeon** [comment]
It looks to me that the code has already been changed so that only ligands in the CIF file are used to look for ligand binding events. This probably means that only those are aligned, but @ConorFWild should probably confirm this.

**Comment by tdudgeon** [comment]
I managed to track down how to add the ligand name to the ligand binding event data. The change could potentially break a lot, but doesn't seem to.
@ConorFWild should review this.
https://github.com/xchem/xchem-align/commit/5183b06906680050f17f1ca8ed462840efa483ec

**Comment by phraenquex** [comment]
@tdudgeon that's terrific news.

Please don't wait for Conor, we can't afford the wait:  go ahead and test it, including with @kaliif ;  and ensure it's thoroughly commented.  

(@ConorFWild can review it later when he looks at #1511.)

**Comment by mwinokan** [comment]
@tdudgeon says that XCA changes for this and #1410 are ready but need loader changes to be able to test thoroughly

e.g. in Jasmin's data there are multiple ligands across multiple chains and the ligand chain assignment determines whether the alignment is correct or not. It's unclear what issues such edge cases could cause downstream in Fragalysis

**Comment by tdudgeon** [comment]
@mwinokan We've spotted a problem with the combisoak data.
In soakDB there's just a single column for the compound code, but there can be multiple compounds and so presumably there should be multiple compound codes.
We need to:
1. work out how this should be handled in soakDB
2. adjust XCA and the loader accordingly.

**Comment by mwinokan** [comment]
Thanks @tdudgeon. I've reached out to Daren

**Comment by kaliif** [comment]
I have the code that loads the combi soak test dataset `combi-soak_3_OCT.tgz`. It assumes the `compound_code` attribute in `meta_aligner.yaml` to be in the same dictionary as `smiles` (as per discussion with Tim).

There are no changes to the schema, the data in the database is stored exactly the same way and the new data will not break anything that's already loaded, however the loader is not compatible with the old data anymore. If merged to staging or prod, all new metadata files must be in the new format. @mwinokan, @tdudgeon how would you like to proceed with the testing?

**Comment by mwinokan** [comment]
Thanks @kaliif, how hard would it be to detect and support both metadata formats?

**Comment by kaliif** [comment]
@mwinokan it's quite a bit of work

**Comment by mwinokan** [comment]
**@tdudgeon please create a "staging" XCA deployment at Diamond so that we can test the new loader changes.**

Following this we can align some realistic combi-soak data and test on a dev stack (maybe Kalev's or Tibor), and only then go to staging

@tdudgeon also explains that the production loader will reject any upload from a non-tagged XCA commit/version

**Comment by tdudgeon** [comment]
The second XCA user environment  is ready.
Run this once:
```
git config --global --add safe.directory /dls/science/groups/i04-1/software/xchem-align-staging
```
Run this every time:
```
source /dls/science/groups/i04-1/software/xchem-align-staging/act
conda activate /dls/science/groups/i04-1/software/xchem-align-staging/env_xchem_align
```

Currently the code is from the `combi_soaks` branch with also includes the covalent ligand changes.

**Comment by mwinokan** [comment]
@tdudgeon the order of the ligands in the CIF is not inconsistent because it's done manually at the moment.

Daren and I are working on an XCE fix to make it consistent, so please assume it is

**Comment by tdudgeon** [comment]
@mwinokan OK, great. That makes things **much** simpler.
The remaining thing is to decide what to do when the data is inconsistent
e.g. `num compound codes` != `num smiles` != `num molecules in CIF`.
Is that an error (meaning fail and fix the problem) or a warning resulting in compound codes not be handled.

**Comment by tdudgeon** [comment]
Initial code is ready. This assumes that if there are `n` compounds in the CIF file then the `CompoundCode` column in soakDB will have `n` semi-colon separated compound codes in **exactly** the same order as the compounds appear in the CIF file (see 2 comments above). As an example for `CHIKV_MacB-x1739` which has 3 compounds I manually poked `HUEY;DEWEY;LOUIE` into the `CompoundCode` column for that record, and what we now get in the metadata is this:

```
      ligand_cif:
        file: upload_1/crystallographic_files/CHIKV_MacB-x1739/CHIKV_MacB-x1739.cif
        sha256: d0556ef2b0c3bd065be7d29a3b37698bd1cb099a5122cb27f6fa6a75231a2a98
        source_file: data/lb32633-6/inputs/dls/labxchem/data/lb32633/lb32633-6/processing/analysis/model_building/CHIKV_MacB-x1739/merged.cif
        ligands:
          LG1:
            smiles: Cc1cn[nH]c1-c1ccncc1
            compound_code: HUEY
          LG2:
            smiles: Nc1nnc2ccccn12
            compound_code: DEWEY
          LIG:
            smiles: O=C1NCCN1
            compound_code: LOUIE
```

Still to be decided is how to handle inconsistencies (number if records in `CompoundCode` column being different to the number of molecules in the CIF file - see 1 comment above). Currently this is a warning (and you get no compound codes), but IMHO it should be an error.

**Comment by mwinokan** [comment]
Thanks @tdudgeon. 

I agree with you, any time the user data does not conform to expectation (and we don't have a way to correct it in the code) it should be an error

**Comment by mwinokan** [comment]
@tdudgeon please promote the warning to an error. Eventually we may have to re-connect the compound codes and SMILES explicitly in soakDB

**Comment by mwinokan** [comment]
@phraenquex asks if this ticket has progressed?

As the cluster is live now @tdudgeon please update the staging XCA environment

**Comment by mwinokan** [comment]
@kaliif has tested the loader changes and they are ready to test with a real dataset

**Comment by mwinokan** [comment]
@phraenquex @kaliif why was this moved to "in staging"? I don't see a PR in the backend for it

**Comment by phraenquex** [comment]
@mwinokan moved it back to "move to staging".  @kaliif I guess you have to issue the PR?

**Comment by Waztom** [comment]
@kaliif has had some challenges, with migration conflicts, of new model fields. He says they can be resolved.

**Comment by mwinokan** [comment]
Blindly approving this so we can push to production. When real combi-soak data causes errors down the line, new tickets will be needed
